integration-cli/docker_cli_build_test.go

```diff
@@ -6894,3 +6894,62 @@ func (s *DockerRegistryAuthHtpasswdSuite) TestBuildWithExternalAuth(c *check.C)
 	out, _, err := runCommandWithOutput(buildCmd)
 	c.Assert(err, check.IsNil, check.Commentf(out))
 }
+
+// Test cases in #22036
+func (s *DockerSuite) TestBuildLabelsOverride(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+
+	// Command line option labels will always override
+	name := "scratchy"
+	expected := `{"bar":"from-flag","foo":"from-flag"}`
+	_, err := buildImage(name,
+		`FROM scratch
+                LABEL foo=from-dockerfile`,
+		true, "--label", "foo=from-flag", "--label", "bar=from-flag")
+	c.Assert(err, check.IsNil)
+
+	res := inspectFieldJSON(c, name, "Config.Labels")
+	if res != expected {
+		c.Fatalf("Labels %s, expected %s", res, expected)
+	}
+
+	name = "from"
+	expected = `{"foo":"from-dockerfile"}`
+	_, err = buildImage(name,
+		`FROM scratch
+                LABEL foo from-dockerfile`,
+		true)
+	c.Assert(err, check.IsNil)
+
+	res = inspectFieldJSON(c, name, "Config.Labels")
+	if res != expected {
+		c.Fatalf("Labels %s, expected %s", res, expected)
+	}
+
+	// Command line option label will override even via `FROM`
+	name = "new"
+	expected = `{"bar":"from-dockerfile2","foo":"new"}`
+	_, err = buildImage(name,
+		`FROM from
+                LABEL bar from-dockerfile2`,
+		true, "--label", "foo=new")
+	c.Assert(err, check.IsNil)
+
+	res = inspectFieldJSON(c, name, "Config.Labels")
+	if res != expected {
+		c.Fatalf("Labels %s, expected %s", res, expected)
+	}
+
+	// Command line option labels with only `FROM`
+	name = "scratchy"
+	expected = `{"bar":"from-flag","foo":"from-flag"}`
+	_, err = buildImage(name,
+		`FROM scratch`,
+		true, "--label", "foo=from-flag", "--label", "bar=from-flag")
+	c.Assert(err, check.IsNil)
+
+	res = inspectFieldJSON(c, name, "Config.Labels")
+	if res != expected {
+		c.Fatalf("Labels %s, expected %s", res, expected)
+	}
```

Comment:
Hi @tiborvass @thaJeztah my last comment was incorrect as I actually executed the incorrect command (had a schedule out so didn't check carefully, in hurry).

In Docker 1.11.0, here is the expected behavior:
(NOTE:  As is discussed in #22036, the command line label options will not be in use when the Dockerfile is only from scratch. The command line label options will override when Dockerfile is from another image. The following are the same but may looks like confusing.)
```
$ docker build -t scratchy --label foo= .

Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratch
 ---> 
Step 2 : LABEL bar from-dockerfile
 ---> Running in f4c190a282ed
 ---> 4f8139a0fda2
Removing intermediate container f4c190a282ed
Successfully built 4f8139a0fda2
$ docker inspect --format "{{ json .Config.Labels }}" scratchy

{"bar":"from-dockerfile","foo":""}
(We passed a --label foo=, which is treated as foo="")
```

```
$ docker build -t scratchy --label bar .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratch
 ---> 
Step 2 : LABEL bar from-dockerfile
 ---> Running in 81979aaafc79
 ---> 26ddc83918ea
Removing intermediate container 81979aaafc79
Successfully built 26ddc83918ea
$ docker inspect --format "{{ json .Config.Labels }}" scratchy
{"bar":"from-dockerfile"}

We passed --label bar which is treated as --label bar="", however, 
docker will not override here so no effect.
```

```
$ docker build -t scratchy --label bar .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratch
 ---> 
Step 2 : LABEL bar from-dockerfile
 ---> Running in 32c8cce82af3
 ---> 4d5aaff2569b
Removing intermediate container 32c8cce82af3
Successfully built 4d5aaff2569b

$ docker build -t scratchy2 --label foo .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM scratchy
 ---> 4d5aaff2569b
Step 2 : LABEL bar from-dockerfile2
 ---> Running in 6f4af1e6fdcb
 ---> a17708718028
Removing intermediate container 6f4af1e6fdcb
Successfully built a17708718028
$ docker inspect --format "{{ json .Config.Labels }}" scratchy2
{"bar":"from-dockerfile2","foo":""}
We passed --label foo which is equal to --label foo="", and in this case,
there is an override so we see "foo": ""
```